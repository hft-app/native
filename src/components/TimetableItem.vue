<template>
  <div v-hash-color="lecture.title" class="lecture">
    <div class="title">
      {{ lecture.title }}
    </div>
    <div class="time">
      {{ $d(lecture.start, 'time') }} - {{ $d(lecture.end, 'time') }}
    </div>
    <div class="info">
      <div v-if="lecture.professor" class="professor data icon-prepend icon">
        <fa-icon icon="user-alt" />
        {{ lecture.professor }}
      </div>
      <div v-if="lecture.room" class="room data icon-prepend icon">
        <fa-icon icon="map-marker-alt" />
        {{ lecture.room }}
      </div>
    </div>
  </div>
</template>
<script>
  export default {
    props: {
      lecture: {
        type: Object,
        required: true
      }
    }
  }
</script>
<style lang="scss" scoped>
  .lecture {
    $padding: 1px;
    position: absolute;
    top: $padding;
    left: $padding;
    right: $padding;
    bottom: $padding;
    color: #FFFFFF;
    display: flex;
    flex-direction: column;
    border-radius: 3px;

    .title {
      font-size: 12px;
      padding: 5px 6px 0;
      display: -webkit-box;
      -webkit-line-clamp: 2;
      -webkit-box-orient: vertical;
      overflow: hidden;
      text-overflow: ellipsis;
      word-wrap: break-word;
      font-weight: 500;
    }

    .time {
      font-size: 10px;
      padding: 4px 6px 0;
    }

    .info {
      font-size: 10px;
      display: flex;
      justify-content: space-between;
      padding: 0 6px 5px;
      position: absolute;
      bottom: 0;
      right: 0;
      left: 0;
    }

    .professor {
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }

    .room {
      flex-shrink: 0;
    }

    .data + .data {
      padding-left: 1em;
    }

    .data:only-child {
      margin-left: auto;
    }
  }
</style>
