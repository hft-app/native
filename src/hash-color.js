import Vue from 'vue';

Vue.directive('hash-color', {
  bind(el, {value}) {
    const colors = ['green', 'blue', 'orange', 'red', 'yellow', 'teal', 'purple', 'pink'];
    let hash = 0;
    for(let i = 0; i < value.length; i++) {
      hash += value.charCodeAt(i);
    }
    hash %= colors.length;
    el.classList.add(colors[hash]);
  }
});
