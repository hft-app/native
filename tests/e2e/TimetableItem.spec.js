import {config, shallowMount} from '@vue/test-utils'
import TimetableItem from 'components/TimetableItem.vue';
import 'hash-color'

config.mocks['$d'] = date => date.valueOf();

describe('TimetableItem', () => {
  test('renders correctly', () => {
    const wrapper = shallowMount(TimetableItem, {
      stubs: ['fa-icon'],
      propsData: {
        lecture: {
          title: 'Advanced Stuff',
          room: '2/433',
          professor: 'Müller',
          start: new Date('Mon Feb 10 2020 08:00:00 GMT+0100'),
          end: new Date('Mon Feb 10 2020 09:30:00 GMT+0100'),
        }
      }
    });
    expect(wrapper.element).toMatchSnapshot()
  });

  test('renders correctly without room', () => {
    const wrapper = shallowMount(TimetableItem, {
      stubs: ['fa-icon'],
      propsData: {
        lecture: {
          title: 'Advanced Stuff',
          professor: 'Müller',
          start: new Date('Mon Feb 10 2020 08:00:00 GMT+0100'),
          end: new Date('Mon Feb 10 2020 09:30:00 GMT+0100'),
        }
      }
    });
    expect(wrapper.element).toMatchSnapshot()
  });

  test('renders correctly without professor', () => {
    const wrapper = shallowMount(TimetableItem, {
      stubs: ['fa-icon'],
      propsData: {
        lecture: {
          title: 'Advanced Stuff',
          room: '2/433',
          start: new Date('Mon Feb 10 2020 08:00:00 GMT+0100'),
          end: new Date('Mon Feb 10 2020 09:30:00 GMT+0100'),
        }
      }
    });
    expect(wrapper.element).toMatchSnapshot()
  })
});



