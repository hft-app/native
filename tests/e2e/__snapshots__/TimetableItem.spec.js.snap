// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimetableItem renders correctly 1`] = `
<div
  class="lecture purple"
>
  <div
    class="title"
  >
    
    Advanced Stuff
  
  </div>
   
  <div
    class="time"
  >
    
    1581318000000 - 1581323400000
  
  </div>
   
  <div
    class="info"
  >
    <div
      class="professor data icon-prepend icon"
    >
      <fa-icon-stub
        icon="user-alt"
      />
      
      Müller
    
    </div>
     
    <div
      class="room data icon-prepend icon"
    >
      <fa-icon-stub
        icon="map-marker-alt"
      />
      
      2/433
    
    </div>
  </div>
</div>
`;

exports[`TimetableItem renders correctly without professor 1`] = `
<div
  class="lecture purple"
>
  <div
    class="title"
  >
    
    Advanced Stuff
  
  </div>
   
  <div
    class="time"
  >
    
    1581318000000 - 1581323400000
  
  </div>
   
  <div
    class="info"
  >
    <!---->
     
    <div
      class="room data icon-prepend icon"
    >
      <fa-icon-stub
        icon="map-marker-alt"
      />
      
      2/433
    
    </div>
  </div>
</div>
`;

exports[`TimetableItem renders correctly without room 1`] = `
<div
  class="lecture purple"
>
  <div
    class="title"
  >
    
    Advanced Stuff
  
  </div>
   
  <div
    class="time"
  >
    
    1581318000000 - 1581323400000
  
  </div>
   
  <div
    class="info"
  >
    <div
      class="professor data icon-prepend icon"
    >
      <fa-icon-stub
        icon="user-alt"
      />
      
      Müller
    
    </div>
     
    <!---->
  </div>
</div>
`;
